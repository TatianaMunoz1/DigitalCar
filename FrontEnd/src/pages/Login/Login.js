import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import axios from "axios";
import Inputs from "../../components/Inputs/Inputs";
import Button from "../../components/Button/Button";
import Swal from "sweetalert2";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons";
import "animate.css";
import styles from "./Login.module.css";
import Url from "../../Url";

const Login = (props) => {
  // const history = useHistory();
  const expressions = {
    password: /^.{6,12}$/, // 6 a 12 digitos.
    email: /^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/,
  };
  const [errorMsg, setErrorMsg] = useState("");
  const [showErrorMsg, setShowErrorMsg] = useState(false);
  const [showMessageLogin, setShowMessageLogin] = useState(false);
  const [userData, setUserData] = useState({
    email: "",
    password: "",
  });
  const [userValidation, setUserValidation] = useState({
    email: false,
    password: false,
  });

  const loginMessage = (name, lastName) => {
    const Toast = Swal.mixin({
      toast: true,
      position: "top-end",
      showConfirmButton: false,
      timer: 3000,
      timerProgressBar: true,
      background: "rgba(239, 243, 246, 1)",
      didOpen: (toast) => {
        toast.addEventListener("mouseenter", Swal.stopTimer);
        toast.addEventListener("mouseleave", Swal.resumeTimer);
      },
    });
    Toast.fire({
      icon: "success",
      title: `<h4>Bienvenido ${name} ${lastName}</h4>`,
    });
  };

  useEffect(() => {
    props.renderHeader();
    return () => props.renderHeader();
  }, []);

  useEffect(() => {
    setShowMessageLogin(props.showMessageLogin);
  }, [showMessageLogin]);

  const handlerSubmit = (e) => {
    e.preventDefault();
    const data = {
      email: userData.email,
      password: userData.password,
    };
    axios
      .post(Url+"/login", data)
      .then((response) => {
        if (response.status === 200) {
          props.changeUser(response.data);
          setShowErrorMsg(false);

          loginMessage(response.data.name, response.data.lastName);
          // history.push("/");
        }
      })
      .catch((error) => {
        setShowErrorMsg(true);
        if (error.response.status === 403)
          setErrorMsg(
            "Por favor, vuelva a intentarlo. Sus credenciales son inválidas"
          );
        else
          setErrorMsg(
            "Lamentablemente no ha podido iniciar sesión. Por favor intente más tarde"
          );
      });
  };

  const handleChange = (e) => {
    const target = e.target;
    const value = target.value;
    const valid = target.validState;
    const name = target.name;

    setUserData({
      ...userData,
      [name]: value,
    });
    setUserValidation({
      ...userValidation,
      [name]: valid,
    });
  };

  return (
    <main className={styles.main}>
      {showMessageLogin && (
        <div className={styles.login_request}>
          <FontAwesomeIcon icon={faExclamationCircle} />
          &nbsp; Para realizar una reserva necesitas estar logueado
        </div>
      )}
      <form className={styles.form} action="" onSubmit={handlerSubmit}>
        <h1 className={styles.title}>Iniciar sesión</h1>
        <Inputs
          className={`email`}
          onChange={handleChange}
          value={userData.email}
          type="email"
          label="Email"
          placeholder="correo@correo.com"
          name="email"
          errorLegend="El correo solo puede contener letras, numeros, puntos, guiones y guion bajo."
          regularExpression={expressions.email}
          validState={userValidation}
          setValidState={setUserValidation}
        />
        <Inputs
          className={`password`}
          onChange={handleChange}
          value={userData.password}
          type="password"
          label="Contraseña"
          name="password"
          errorLegend="La contraseña tiene que ser de 6 a 12 dígitos."
          regularExpression={expressions.password}
          validState={userValidation}
          setValidState={setUserValidation}
        />

        {showErrorMsg && <p className={styles.error_message}>{errorMsg}</p>}
        <div className={styles.button_container}>
          <Button
            type="submit"
            title="Enviar"
            btnStyle="btn--full"
            btnSize="btn--responsive"
            onClick={handlerSubmit}
          />
          <p className={styles.register_box}>
            ¿Aún no tienes una cuenta? &nbsp;
            <Link to="/register">
              <span className={styles.register_label}>Regístrate</span>
            </Link>
          </p>
        </div>
      </form>
    </main>
  );
};

export default Login;
