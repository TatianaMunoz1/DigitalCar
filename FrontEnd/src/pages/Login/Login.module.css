.hidden {
  display: none;
}
.main {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: white;
  background-color: black;
  height: 100vh;
  justify-content: center;
}
.title {
  font-family: "Roboto";
  font-size: 25px;
  padding-top: 0px;
}
.form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 10px 0px 0px 0px;
  gap: 15px;
  width: 300px;
}
.button_container {
  padding: 10px 0px 0px 0px;
  display: flex;
  flex-direction: column;
  width: 100%;
}
.register_box {
  padding-top: 10px;
  font-size: 15px;
  align-self: center;
}
.register_label {
  color: #ff7e19;
  text-decoration-color: #ff7e19;
  font-weight: bold;
}
.login_request {
  color: #b00020;
  background: #ffcfcf;
  height: auto;
  width: 300px;
  font-size: 14px;
  line-height: 20px;
  padding: 15px;
  margin-bottom: 15px;
  border-radius: 8px;
}
.error_message {
  color: red;
  width: 300px;
  text-align: center;
}

/* -----TABLET----- */
@media (min-width: 760px) and (max-width: 1280px) {
  .form {
    width: 350px;
    gap: 30px;
  }
  .title {
    padding-bottom: 20px;
  }
  .error_message {
    width: 450px;
  }
  .login_request {
    width: 350px;
    padding: 20px;
    margin-bottom: 25px;
  }
  .button_container {
    padding: 20px 0px 0px 0px;
  }
}

/* -----DESKTOPS----- */
@media (min-width: 1281px) {
  .form {
    width: 450px;
    gap: 35px;
  }
  .title {
    font-size: 1.7rem;
  }
  .login_request {
    width: 450px;
    padding: 20px;
    margin-bottom: 25px;
  }
  .error_message {
    width: 450px;
  }
  .button_container {
    padding: 20px 0px 0px 0px;
  }
}
