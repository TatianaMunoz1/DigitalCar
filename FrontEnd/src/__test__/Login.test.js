import React from "react";
import '@testing-library/jest-dom'
import { shallow } from "enzyme";
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Login from "../pages/Login/Login.js"


describe ('<Login />',() => {
    /*const componenteInput =
    ({ email, password, onChange }) => (
        <div>
          <input name="email" value={email} onChange={onChange} />
          <input name="password" value={password} onChange={onChange} />
        </div>
      );
    componenteInput.propTypes={
        email: PropTypes.email.isRequired,
        password: PropTypes.password.isRequired,
        onChange: PropTypes.func.isRequired,
    };
    
    const testState = { email:"correo@correo.com" , password: "digitalcar123" };
        const wrapper = shallow((
  <Login
    email={testState.email}
    password={testState.password}
    onChange={(e) => {
      testState[e.target.name] = e.target.value;
    }}
  />
  ));
expect(wrapper.find('input').at(0).prop('value')).toEqual("correo@correo.com");
expect(wrapper.find('input').at(1).prop('value')).toEqual("digitalcar123");
wrapper.find('input').at(0).simulate('change', { target: { name: 'email', value: "digitalcar@correo.com" } });
wrapper.find('input').at(1).simulate('change', { target: { name: 'password', value: "digitalcar4321" } });
expect(testState.email).toEqual("digitalcar@correo.com");
expect(testState.password).toEqual("digitalcar4321");*/

    const componenteInput = [{
        id: "testemail",
        name :"testemail",
        placeholder: "correo@correo.com",
        type : "email",
        value : "digitalcar@correo.com"
      },
     {
        id: "testpasword",
        name :"testpasword",
        placeholder: "digitalcar123",
        type : "password",
        value : "digitalcar4321"
      }]
    test("render add features ", () => {
        const wrapper = shallow(<Login/>)
        expect(wrapper).toMatchSnapshot();
    });
    
    test('email check',()=>{
        const onChangeSpy = jest.fn();
        const wrapper = shallow(<Login componente={componenteInput} handleChange ={onChangeSpy}/>);
        wrapper.find('.email').simulate('change', {target: 
            {name: 'email', value: "digitalcar@correo.com"}});
        expect(wrapper.state('email')).toEqual('digitalcar@correo.com');
    })

    test('password check',()=>{
        const wrapper = shallow(<Login componente={componenteInput}/>);
        wrapper.find('Inputs [type="password"]').simulate('change', {target: {name: 'password', value: "digitalcar123"}});
        expect(wrapper.state('password')).toEqual('digitalcar123');
    });

});