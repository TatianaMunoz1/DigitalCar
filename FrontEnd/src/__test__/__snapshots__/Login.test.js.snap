// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> render add features  1`] = `ShallowWrapper {}`;
